rfcs/hse/0008-rework-cn-stats.md

```diff
@@ -0,0 +1,238 @@
+   measure of on-media key length.
+
+## Solution
+
+### Overview
+
+1. Reduce the number of stat struct types.
+1. Simplify on-media storage of stats.
+1. Eliminate [Misleading statistics](#Misleading-statistics).
+1. Eliminate the need to compute kvset level stats by iterating over all kblocks in kvset.
+
+### Details
+
+1. Eliminate `struct kblk_metrics` and `struct kvset_metrics` use `struct kvset_stats` more widely.
+   - Advantages:
+     - Eliminate the overhead of computing `struct kvset_metrics`.
+
+1. In `struct kvset_stats`, store `vlen` and `vcount` stats for each `enum
+   kmd_vtype` and eliminate special named fields such as `nptombs`.
+   - Advantages:
+     - Eliminates all problems identified in [Misleading statistics](#Misleading-statistics).
+
+1. Store kvset-specific instances of the `one_stat_struct` in kvset hblocks
+   and remove related data them from CNDB.
+   - Advantages:
+     - Would not have to walk all kblocks to compute kvset level stats.
+     - Simplify CNDB.
+     - Simplify kvset open process.
+
+1. Store kblock-specific members of `struct kvset_stats` in kblock headers for
+   use during node split, but provide access to it in memory as an instance of
+   `struct kvset_stats`.
+   - Advantages:
+     - Eliminate need for `struct kblk_metrics`
+     - Dead simple to compute kvset stats from kblock stats.
+
+1. Use `struct kvset_stats` with kvset builders instead of `struct key_stats`.
+   - Advantages:
+     - Eliminate need for `struct key_stats`.
+
+1. Replace `vused` with its inverse, `vgarbage`
+   - Advantages:
+     - A kv-compacted kvset will have `vgarbage == 0`, which is much easier to
+       work with than `vused` which has an unspecified non-zero value for nicely
+       compacted kvset (it equals the vblock size minus overhead such as headers
+       and padding).
+
+#### Proposed data structure to replace kvset_metrics and kvset_stats
+
+The following definitions will replace `struct kvset_metrics` and `struct kvset_stats`.
+
+```
+/* Stats about items stored on-media
+ *
+ * Statistics about keys, values, kblocks, etc. all of which are stored
+ * on-media.
+ *
+ * Notes on omlen vs ulen:
+ * - Member 'omlen' refers the amount of storage used to hold the items, while
+ *   'ulen' refers the size of the items from the user's perspective.
+ * - For items such as keys and values, 'omlen' will be less than 'ulen' due to
+ *   longest common prefix elimination (keys) and compression (values).
+ * - For items such as hblocks and kblocks, 'omlen' will greater than 'ulen' due
+ *   to padding and allocation overhead.
+ */
+struct om_stats {
+    uint64_t count;  // Number of items on media
+    uint64_t omlen;  // On-media length of items, in bytes
+    uint64_t ulen;   // User length of items, in bytes
+};
+
+/* Key and value statistics.
+ *
+ * This struct is used for kvset, node and tree stats.
+ */
+struct kvstats {
+    struct om_stats vtype[NUM_KMD_VTYPES]; // on-media stats for values by vtype
```

Comment:
Maybe change this constant to VTYPES_COUNT per my comment in the implementing PR?